spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/PriorityHashMap.scala

```diff
@@ -0,0 +1,330 @@
+ * - an array storing indexes into the first two arrays, used as an inline hash-table allowing to
+ *   quickly locate keys in the heap in constant time.
+ *
+ * The indexes hash-table doesn't use overflow lists for dealing with hash collisions.
+ * The overflow entries are placed in the main hash-table array in the first not-taken
+ * entry to the right from the original position pointed by key hash. On search,
+ * if the key is not found immediately at a position pointed by key hash, it is searched
+ * to the right, until it is found or an empty array entry is found.
+ *
+ * @param _capacity minimum required capacity of this collection; the actual capacity may be larger than this,
+ *                  because for performance reasons it is rounded up to the nearest power of two
+ * @tparam K type of keys
+ * @tparam V type of values; values must be comparable
+ */
+class PriorityHashMap[K, V : Ordering](_capacity: Int) {
+
+  private[this] var _size = 0
+  def size = _size
+  def isEmpty = _size == 0
+  def nonEmpty = !isEmpty
+
+  private def log2(n: Int): Int = 32 - Integer.numberOfLeadingZeros(n - 1)
+  private def pow2(n: Int): Int = math.pow(2, n).toInt
+
+  /** The maximum number of items that can be stored at a time in this map. */
+  val capacity = pow2(log2(math.max(_capacity, 2)))
+
+  // indexes array is twice bigger than capacity,
+  // so we need to use twice bigger hashing mask as well;
+  // the mask allows to filter appropriate number of less significant bits of the hash
+  private val mask = (capacity * 2) - 1
+
+  /** Original hash multiplied by 2, to hash into even entries only, so there is
+    * initially at least one empty entry between them (works as a search-terminator).
+    * This is to protect against bad hashes forming long sequences
+    * of consecutive numbers, which would result in O(n) lookup, instead of O(1), even
+    * if there were no hash-collisions. */
+  private def hash(key: K): Int =
+    (key.hashCode() << 1) & mask
+
+  private[this] val _keys = Array.ofDim[AnyRef](capacity).asInstanceOf[Array[K]]
+  private[this] val _values = Array.ofDim[AnyRef](capacity).asInstanceOf[Array[V]]
+  private[this] val _indexes = Array.fill(capacity * 2)(-1)  // hash-table twice as big, to account for hash-collisions
+  private[this] val ordering = implicitly[Ordering[V]]
+
+  /** Finds a key in the indexes array.
+    * Returns a position in the indexes array pointing to the found key or a position of
+    * the first empty index entry, if key was not found. */
+  private def find(key: K): Int = {
+    find(key, hash(key))
+  }
+
+  /** Finds a key in the indexes array, starting at a given position in the indexes array
+    * Returns a position in the indexes array pointing to the found key or a position of
+    * the first empty index entry, if key was not found. */
+  @tailrec
+  private def find(key: K, pos: Int): Int = {
+    val i = _indexes(pos)
+    if (i < 0 || _keys(i) == key) pos
```

Comment:
It would only make sense if we memorized key hashes. Otherwise, computing hashes for both keys is probably just as costly as comparing objects directly.